clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -738,16 +738,47 @@ Parser::DeclGroupPtrTy Parser::ParseUsingDeclaration(
       return nullptr;
     }
 
-    if (!Tok.is(tok::identifier)) {
+    Decl *UED = nullptr;
+
+    if (Tok.is(tok::identifier)) {
```

Comment:
This assumes that only 2 scenario will occur:

* Either there is an already annotated template
* Or the identifier is not part of a simple-template-id

I am not sure that is correct.
ie there is scenario where we find a `tok::less` after the identifier.

I'd like to understand exactly what happens, in particular:

* Whether that scenario can ever happen (and we should assert otherwise)
* What's the difference between this code and `ParseBaseTypeSpecifier`: Either we can remove a lot of code from  `ParseBaseTypeSpecifier`, or factorize both functions

Afaict, we always get an annotation because `ParseOptionalCXXScopeSpecifier` annotates.
but that should be the case in `ParseBaseTypeSpecifier` too/

Reply:
Line 723, `ParseOptionalCXXScopeSpecifier`, there are options to make it a type only context for error recovery (and `Conttext` is misspelled.

Reply:
1) You're correct that `ParseOptionalCXXScopeSpecifier` is going to provide us a `TemplateIdAnnotation` every time there is an identifier followed by a "less than" sign.

2) On top of what I do here, `ParseBaseTypeSpecifier` handles `computed-type-specifier` (which is `decltype` and pack indexing) and dependent types. If you take a look at which tokens that function expects, they are `tok::kw_decltype`, `tok::annot_decltype`, `tok::annot_pack_indexing_type`, `tok::annot_template_id`, and `tok::identifier`. If we remove tokens that are coming from `computed-type-specifier`, we'll be left with just `tok::annot_template_id` and `tok::identifier`, which are exactly the tokens I expect.

> Line 723, ParseOptionalCXXScopeSpecifier, there are options to make it a type only context for error recovery

I'd be wary to mark this as a type-only context, unless the Standard says so. I believe it doesn't for `using enum`, but does for base specifier, because the latter uses `class-or-decltype` grammar.

Reply:
After more tests, I realize that the handling of `less` in `ParseBaseTypeSpecifier` is for error recovery, and we might want to do the same thing.
https://godbolt.org/z/9d3Pvxjxe

I agree that we should not factorize out decltype/pack indexing, but everything else is hopefully exactly the
same.

> I'd be wary to mark this as a type-only context, unless the Standard says so. I believe it doesn't for using enum, but does for base specifier, because the latter uses class-or-decltype grammar.

Sorry, I was not clear. We always expect a type here in the grammar (ie, `using enum expr` is never valid), and setting `IsTypename` only improves error recovery (a bit).

(type-only context in the specification refers specifically to places where `typename` can be ellided, and typename is never valid here)

Reply:
We agreed offline that adding FIXME would be enough for now.